psyneulink/core/llvm/execution.py

```diff
@@ -567,6 +562,16 @@ def _initialize_autodiff_struct(self, autodiff_stimuli, arg_index,
 
 
     def run(self, inputs, runs=0, num_input_sets=0, learning=False):
+        if isgenerator(inputs):
+            list_inputs = list(inputs)
+            inputs = {k:[list(v)] for (k,v) in list_inputs[0].items()}
+            for i, inp in enumerate(list_inputs):
+                if i == 0:
+                    continue
+                for k,v in inp.items():
+                    inputs[k].append(v)
+            num_input_sets = len(next(iter(inputs.values())))
+            runs = num_input_sets
```

Comment:
Is there another way to handle generators without converting them to lists for compiled execution?

Reply:
I think it might be better to leave generators unsupported and let people use `LLVMExec`. The main advantage of generators is that not all values are in memory, and that is eliminated by creating a list.
If you want to support generators it'd be better to create a ctype structure directly.
The added code converts generator (list) of input sets into a dictionary of inputs for each input ndoe, which is then converted back to a list of single input per input node in `_get_run_input_struct`.

The second part was wrt `num_input_sets` isn't  the value the same as `len(list_inputs)`?

Reply:
ok I will look into replacing the generator code with a compiled equivalent of `compositionrunner` - for now, consider it a hack (will mark it as such). The main reason I went for `LLVMRun` instead of `LLVMExec` here was to get the additional gains from loop code-motion optimizations (as it previously did by running all epochs in a loop). I should test to see if they make a huge performance difference though.

I wasn't exactly sure what `num_input_sets` referred to - is it the same as the number of trials in the input?

Reply:
It'd be good to support both `LLVMExec` and `LLVMRun`. `LLVMRun` usually does perform better, but it depends on how many trials you want to run and how costly it is to set up. Going back and forth between nodes first vs. inputs first representation might take more time than `LLVMExec`.

`num_input_sets` is just a number of inputs. You can run more trials than `num_input_sets` in which case the run will loop over and start reusing inputs from the start. I'm not sure how it's supposed to work with generators, I'd expect that in this case `num_trials` and `num_input_sets` need to match.

It's not clear to me what `compositionrunner` should do. At a high level it looks like it's doing the same/similar job as a composition controller.

Reply:
`compositionrunner` is respnsible for batching inputs for learning cases. It also handles things like patience and epoch count.
The reasoning behind keeping it in a separate class was to isolate the compiled code, and also keep the methods specific to it from polluting `composition.py` (which is already quite a massive file)